simpleviewinc/keg-hub
# ReTheme Hook Updates

**Ticket**: [ZEN-467](https://jira.simpleviewtools.com/browse/ZEN-467)

[Semver-Status] minor

### Updates - 11/30/20
* Removed the Touchable work-around
* Fixed tests for re-theme
* Added JSDoc notes where needed
* Pushed new docker containers

**Notes**
* These changes mean you will not see much of a difference from the different touchable states
* One thing you will see is,
  * When you hover over a button
  * Then click, and hold down the mouse button
  * Then drag off the button, and release the mouse button
    * In the past this caused the button to stay in a hovered state
    * It should now revert back to the default state as expected
* To check that the hooks states are working, outside the touchable, use the retheme docs app
  * Toggle open the hooks section and test that each hook works as expected


## Context

* When we recently upgrade to rnw v14, it introduced a new way for hover to work
* One approach is to use a new `Pressable` component that passes the hover boolean to child components
  * Unfortunately This introduces another dom-element that could interfere with styling.
  * We should update the `useThemeHover` hook to work with rnw 14, ensuring all our existing components will work without needing changes and without `Pressable` interfering with styling

## Goal
  * Fix the retheme hooks so the work with the updated version of `react-native-web`
  * Update the hooks in a way so the ReTheme consumers do not need to change their code
  * Ensure the hooks work without adding extra dom-nodes

## Updates

**Re-Theme**
* `repos/re-theme/app/package.json`
  * Updated dependencies for the app
* `repos/re-theme/configs/jest.config.js
  * Moved out of the package.json
* `repos/re-theme/configs/buildHook.js`
  * Cleaned up the build hook to only run once
* `repos/re-theme/configs/rollup.config.js`
  * Cleaned up alias to be reusable
    * Now match how keg-components alias work
  * Updated react-native-web imports to use the esm imports
  * The CJS imports were being included in the bundle, so the code was being duplicated
    * These exports are now considered external, and will be imported from the consumers version of react-native-web
* `repos/re-theme/src/styleInjector/reactNativeWeb.js`
  * Updated to use the `esm` imports instead of the `cjs` versions
  * This allows the imports to be external, and not be included in the bundle
  * This way the `react-native-code` that's imported in not duplicated
* `repos/re-theme/container/*`
  * Updated the container files to use the `KEG_PROXY_PORT` for registering with the `keg-proxy`
* `repos/re-theme/scripts/dockerBuild.sh`
  * Added a docker build script for building within the docker container
* `repos/re-theme/src/hooks/*`
  * Rewrote how the theme hooks work to make them work with the newer versions of react-native-web
    * They now use the dom element to apply the event handlers
    * This should allow them to be more maintainable long term
    * So we should not have to update them again if things change within `react-native-web`
* `repos/re-theme/src/hooks/usePointerState.js`
  * Added a new pointerState hook, that allows tracking the current state of a pointer ( mouse )
  * This is used by the `useThemeHook`, but also exported so it can be used in other places
    * Allows tracking `hover`, `active`, or `focus` states
* `repos/re-theme/src/hooks/hookFactory.js`
  * Removed some files no longer being used
* `repos/re-theme/src/hooks/nativeThemeHook.js`
  * Updated to use shallowEqual because it's faster
* `repos/re-theme/src/hooks/__tests__/*`
  * Added more tests for updated code
  * Fixed test the broke due to code changes
* `repos/re-theme/src/mocks/reactHooks.js`
  * Added mocks for react hooks so they can be reused
  * I eventually want to move things like this in to a `jest` helper repo
  * But that's down the road, and this works for now

**Keg-Components**
* `repos/keg-components/configs/aliases.json`
  * Added alias for the `useClassOnPressIn` hook
* `repos/keg-components/src/components/touchable/touchable.native.js`
  * Added `useClassOnPressIn` hook to work around class issues
* `repos/keg-components/src/hooks/useClassOnPressIn/useClassOnPressIn.js`
  * Hook to work around class issue for the Touchable component

## Testing

**Test 1**
* Run the retheme docker package => `keg retheme pack run docker.pkg.github.com/simpleviewinc/keg-packages/retheme:zen-467-retheme-hook-updates`
  * Navigate to http://retheme-zen-467-retheme-hook-updates.local.kegdev.xyz/
  * Open each of the hooks, and ensure they work as expexted

**Test 2**
* Attach to the running docker container => `keg d att package-retheme-zen-467-retheme-hook-updates`
* Run the tests => `yarn test`
  * Ensure all tests pass
 
**Test 3**
* Run the keg-components docker package => `keg dpg run docker.pkg.github.com/simpleviewinc/keg-packages/keg-components:zen-467-retheme-hook-updates`
* Navigate to http://components-retheme-hook-updates.local.kegdev.xyz
* Ensure everything works as expected
  * Test out button hovers
  * Test clicking and holding down a button, then dragging the mouse off the button. Then letting go
   * The main button issues including the disappearing / style issues should be resolved
    * I tried every possible thing to try and recreate the issue, and could not, but it's possible it still happens
      * Basically the Button is losing its custom classes in some cases, which causes it to disappear, or flicker
      * I added a work around that ensure the classes are applied when the button is clicked
         * See the `useClassOnPressIn` hook added to the `Touchable component`
        * This works for now, but a proper workaround will be added with the  [Pressable Ticket](https://jira.simpleviewtools.com/browse/ZEN-468)

**Test 4**
* Run the events-force docker package => `keg evf pack run docker.pkg.github.com/simpleviewinc/keg-packages/tap:retheme-hook-updates`
* Navigate to http://evf-retheme-hook-updates.local.kegdev.xyz
* Ensure everything works as expected
  * Test out button hovers
  * Test clicking and holding down a button, then dragging the mouse off the button. Then letting go
   * The main button issues including the disappearing / style issues should be resolved
    * I tried every possible thing to try and recreate the issue, and could not, but it's possible it still happens
      * Basically the Button is losing its custom classes in some cases, which causes it to disappear, or flicker
      * I added a work around that ensure the classes are applied when the button is clicked
         * See the `useClassOnPressIn` hook added to the `Touchable component`
        * This works for now, but a proper workaround will be added with the  [Pressable Ticket](https://jira.simpleviewtools.com/browse/ZEN-468)
